### PR TITLE
Build image on eLife's infrastructure, run tests

### PR DESCRIPTION
**Pull request by giorgiosironi**
The first step of https://github.com/elifesciences/elife-xpub/issues/675

Ports over the following Gitlab CI jobs, using a `docker-compose` configuration:
- `build`
- `lint`
- `test`
- `test:browser`
- `test:dependencies`

These provide feedback in the form of commit status checks. Execution time seems slightly better (~20-30 seconds less per job) than on Gitlab CI, probably because of lower overhead when running on a single node but possibly also dedicated hardware.

There is more in Gitlab CI, but some of these could be reasonably removed from there once they are stable on this side, reducing the execution time.

### CODE REVIEW

File: .env.ci [hunk 1]
```diff
@@ -0,0 +1,16 @@
+COMPOSE_PROJECT_NAME=elife-xpub
+IMAGE_TAG=develop
+PORT=3000
+
+# connection details for tests
+# TODO: confusing, should just be PGHOST?
+PGHOSTADDR=postgres
+PGHOST=postgres
```
Comment (tamlyn) [inline .env.ci]: `PGHOSTADDR` was added in June last year a commit entitled `fix(ci): change PG_HOST for k8s`. I reckon it should be safe to remove it here.

File: Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+                        sh "IMAGE_TAG=${commit} docker-compose -f docker-compose.ci.yml run --rm --name elife-xpub_app_lint app npm run lint"
+                    }, 'lint', commit)
+                },
+                'test': {
+                    withCommitStatus({
+                        sh "IMAGE_TAG=${commit} docker-compose -f docker-compose.ci.yml run --rm --name elife-xpub_app_test app npm test"
+                    }, 'test', commit)
+                },
+                // TODO: not sure this can run in parallel with `test`?
+                'test:browser': {
```
[the remark this answers is not in the document]
Comment (giorgiosironi) [reply Jenkinsfile]: They run with separate `docker-compose run` commands, but there is only a single `postgres` service that they would depend on; hence there is the potential for a race condition between the two.

**Comment by giorgiosironi** [comment]
Trying to separate `test` and `test:browser` databases, will have to debug a bit.

**Comment by tamlyn** [comment]
In this case `PGDATABASE` environment variable is ignored because 'config/test.js` specifies the database should be `test`. In `node-config` environment variables trump config files but confusingly these environment variables are not used by `node-config`, they are used directly by `node-pg`. So need to either add the `PG*` variables to `test/custom-environment-variables.js` or specify the database some other way.

**Comment by giorgiosironi** [comment]
I don't have an explanation yet, but it seems `test:browser` was trying to use the separate database:
```
12:00:38 [test:browser]       error: database "test_browser" does not exist
```
but this wasn't initialized, so I was initially looking into a problem with the `postgres` container setup.

**Comment by giorgiosironi** [comment]
Re-running the command manually and entering the `postgres` command shows a populated `test_browser` database:
```
test_browser=# \dt
           List of relations
 Schema |     Name     | Type  | Owner 
--------+--------------+-------+-------
 public | audit_log    | table | test
 public | ejp_name     | table | test
 public | entities     | table | test
 public | file         | table | test
 public | identity     | table | test
 public | journal      | table | test
 public | manuscript   | table | test
 public | migrations   | table | test
 public | organization | table | test
 public | review       | table | test
 public | team         | table | test
 public | user         | table | test
(12 rows)
```
and the tests pass:
```
 ✓ Happy path
 ✓ Ability to progress through the wizard is tied to validation
 ✓ Form entries are saved when a user navigates to the next page of the wizard
```

However, `postgres` may take too long to create the database; will add waits to make sure there's no race on that.

File: Jenkinsfile [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+                    try {
+                        withCommitStatus({
+                            sh "IMAGE_TAG=${commit} NODE_ENV=production NODE_CONFIG_ENV=test PGDATABASE=test_browser docker-compose -f docker-compose.ci.yml run --rm --name elife-xpub_app_test_browser app bash -c './scripts/wait-for-it.sh postgres:5432 && npm run test:browser -- --screenshots /tmp/screenshots --screenshots-on-fails'"
+                        }, 'test:browser', commit)
+                    } finally {
+                        archiveArtifacts artifacts: "build/screenshots/*", allowEmptyArchive: true
```
[the remark this answers is not in the document]
Comment (giorgiosironi) [reply Jenkinsfile]: If I understood that right, it's `./build/screenshots/` on the host mounted into the container at the `/tmp/screenshot` path; is that correct?

[the remark this answers is not in the document]
Comment (damnhipster) [reply Jenkinsfile]: Correct

**Comment by giorgiosironi** [comment]
Ok, green again, testing robustness a couple of times, and merging.

**Pull request merged**